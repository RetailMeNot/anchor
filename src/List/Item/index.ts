export { Item } from './Item.component';
