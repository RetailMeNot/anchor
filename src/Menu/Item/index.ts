export { default as Item } from './Item.component';
