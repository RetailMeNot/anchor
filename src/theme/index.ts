export { default as colors } from './colors.theme';
export { default as fonts } from './fonts.theme';
export { default as sizes } from './sizes.theme';
export { default as variables } from './variables.theme';

export { NormalizeCSS, GlobalCSS } from './GlobalStyles';
