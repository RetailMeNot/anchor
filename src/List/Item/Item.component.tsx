// REACT
import * as React from 'react';
// VENDOR
import classNames from 'classnames';
import styled from 'styled-components';
// COMPONENTS
// THEME
import { colors, fonts, sizes, variables } from '../../theme';

export interface ItemProps {
  children?: any;
  className?: string;
}

export const StyledItem = styled.a`
  display: flex;
  border-radius: ${variables.borderRadius};
  padding: ${sizes.padding.sm} ${sizes.padding.md};
  cursor: pointer;
  color: ${colors.neutrals.charcoal.base};
  font-family: ${fonts.fontFamily};
  font-weight: normal;
  text-decoration: none;
  line-height: 1.5rem;
  transition: background-color 500ms;

  &:hover {
    background-color: rgba(0, 0, 0, .05);
  }
`;

const DefaultProps: ItemProps = {
};

export const Item = ({
  className,
  children,
  ...props
}: ItemProps = DefaultProps): React.ReactElement<any> => (
  <StyledItem className={classNames(className)} {...props}>
    {children}
  </StyledItem>
);

Item.defaultProps = DefaultProps;

export default Item;
