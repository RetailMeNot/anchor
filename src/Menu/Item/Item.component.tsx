// REACT
import * as React from 'react';
// VENDOR
import classNames from 'classnames';
import styled from 'styled-components';
// COMPONENTS
// THEME
import { colors } from '../../theme';

export interface ItemProps {
  path: string;
  label?: string;
  children?: any;
  className?: string;
  active?: boolean;
  // TODO: do I need this?
  size?: 'small' | 'large';
  // TODO: as to allow element overrides
  as?: any;
  type?: any;
}

export const StyledItem = styled.a`
  cursor: pointer;
  text-align: center;
  color: ${colors.neutrals.silver.dark};
  text-decoration: none;
  transition: color 250ms;

  // Disable Select
  user-select: none;
  -webkit-touch-callout: none;

  // TODO: this doesn't go here (Law of Demeter)
  /* Icons */
  svg {
    max-width: 8px;
  }
`;

const DefaultProps: ItemProps = {
  path: '/',
  as: 'a',
};

export const Item = ({
  className,
  children,
  label,
  size,
  active,
  ...props
}: ItemProps = DefaultProps): React.ReactElement<any> => (
  <StyledItem className={classNames(className, { active })} {...props}>
    {label ? label : children}
  </StyledItem>
);

Item.defaultProps = DefaultProps;

export default Item;
