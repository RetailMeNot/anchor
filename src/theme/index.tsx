// Theme Files
import { typography } from './typography.theme';
import { ColorsTheme } from './colors.theme';
import { ThemeProvider as XstyledThemeProvider } from '@xstyled/styled-components';
import styled from '@xstyled/styled-components';
import { space as spaceStyles, SpaceProps } from '@xstyled/system';
import { BUTTON_KEY, BUTTON_THEME } from '../Button';
import { INPUT_KEY, INPUT_THEME } from '../Form/Input/utils';
import React from 'react';
import {Button} from '../Button';
import classNames from 'classnames';

export const RootTheme = {
    typography,
    radii: { none: '0', base: '4px', modal: '8px', circular: '1000px' },
    [BUTTON_KEY]: BUTTON_THEME,
    [INPUT_KEY]: INPUT_THEME,
    colors: ColorsTheme,
    fonts: {
        base:
            'Avenir Next, Segoe UI, Helvetica Neue, Helvetica, Roboto, sans-serif',
        mono: `'SFMono-Regular', Consolas, Liberation Mono, Menlo, Courier, monospace`,
    },
    // These are all minimums
    breakpoints: {
        xs: 0,
        sm: 576,
        md: 768,
        lg: 992,
        xl: 1200,
        xxl: 1600,
    },
    // todo: we should likely remove the color.borders colors and just use this
    // or find a way to reference the colors.border colors here
    // e.g. `thin solid borders.base`
    borders: {
        light: `thin solid ${ColorsTheme.borders.light}`,
        base: `thin solid ${ColorsTheme.borders.base}`,
        dark: `thin solid ${ColorsTheme.borders.dark}`,
    },

    skeleton: {
        colorStart: '#E7E7E7',
        colorEnd: '#D3D3D3',
    },
};

// Exporting out @xstyled's ThemeProvider. Only way I could get it to work w/o issues.
// export class ThemeProvider2 extends XstyledThemeProvider {}

interface IterableObject {
    [s: number]: number | string | boolean | IterableObject;
  }

interface DebugProps {
    property: number | string | boolean | IterableObject;
    propertyName: string;
    rootProperty?: boolean;
}

interface StyledDebugProps {
    indent?: number;
}

const StyledDebug = styled('div')<StyledDebugProps>`
    font-family: mono;
    margin-top: 0.5rem;
    color: white;

    .category .category h2 {
        background-color: #ec8947;
    }

    .category .category .box {
        border-color: #ec8947;
    }

    .category .category .category h2 {
        background-color: #b84343;
        color: white;
    }

    .category .category .category .box {
        border-color: #b84343;
    }

    .category .category .category .category h2 {
        background-color: #91cdd2;
        color: black;
    }

    .category .category .category .category .box {
        border-color: #91cdd2;
    }

    .category .category .category .category .category h2 {
        background-color: #567224;
        color: white;
    }

    .category .category .category .category .category .box {
        border-color: #567224;
    }
`;

const StyledH2 = styled('h2')`
    ${spaceStyles}
    background-color: #ffea92;
    color: black;
    margin-bottom: 0;
    margin-top: 0;
    padding: 0.25rem 0.5rem 0.25rem 1rem;
    font-size: 1rem;
    cursor: pointer;
    border-radius: 0 1rem 1rem 0;
`;

interface BoxProps extends SpaceProps {
    show?: boolean;
}

const Box = styled('div')<BoxProps>`
    ${spaceStyles}
    width: 20rem;
    border-left: solid #ffea92 2px;
    margin: 0.5rem;
    ${({show}) => `display: ${show ? 'block' : 'none'}`}
`;

const Category = styled('div')<SpaceProps>`
    ${spaceStyles}
    width: 20rem;
`;


const ColorBox = styled('div')<any>`
    width: 1rem;
    height: 1rem;
    display: inline-block;
    vertical-align: middle;
    margin-left: 0.5rem;
    background-color: ${({backgroundColor}) => backgroundColor};
    border: solid white 1px;
`;

const Debug = ({property, propertyName}: DebugProps) => {
    const [show, setShow] = React.useState(false);

    return(
        <StyledDebug>
            {(property || property === 0)
                ? (
                    typeof property === 'number' ||
                    typeof property === 'string' ||
                    typeof property === 'boolean' ||
                    typeof property === 'function'
                        ? (
                                <>
                                    <span style={{fontWeight: 'bold'}}>{propertyName}: </span>
                                    <span>
                                        {typeof property === 'function'
                                            ? 'function'
                                            : /^#[a-zA-Z0-9]{4,7}/.test(property.toString())
                                                ? (<>
                                                    <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                                    <ColorBox backgroundColor={property} />
                                                    </>)
                                                : <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                        }
                                    </span>
                                </>
                            )
                        :
                            <Category pl={0} className="category">
                                <StyledH2 onClick={() => setShow(!show)}>{propertyName}</StyledH2>
                                <Box pl={2} className="box" show={show}>
                                    {Object.values(property).map((_property, index) => (
                                        <Debug
                                            key={index}
                                            property={_property}
                                            propertyName={Object.getOwnPropertyNames(property)[index]}
                                        />
                                    ))}
                                </Box>
                            </Category>
                    )
                : (
                    <>
                       Empty.
                    </>
                )
            }
        </StyledDebug>
    );
};

const DebugWrapper = styled('div')`
    background-color: rgba(0,0,0,.75);
    border-left-color: transparent;
    position: fixed;
    left: 0;
    top: 0;
    height: 100%;
    overflow: hidden;
    width: 30rem;
    padding: 1rem 2rem 1rem 0;

    &.closed {
        height:2.4rem;
        width: auto;
        padding-right: 1rem;

        .debug-scroll {
            display: none;
        }
    }
`;

const DebugScroll = styled('div')`
    height: 100%;
    overflow-y: auto;
    width: 100%;
`;

interface ThemeProviderProps {
    children?: any;
    debug?: boolean;
    theme?: any;
}

export const ThemeProvider = ({ children, debug, theme }: ThemeProviderProps) => {
    const isOpen = localStorage.themeProviderDebug === 'true';
    const [isOpenState, setIsOpenState] = React.useState(isOpen);

    return (
        <XstyledThemeProvider theme={theme}>
            {debug &&
                <DebugWrapper className={classNames(!isOpenState ? 'closed' : null)}>
                    <Button
                        onClick={() => {
                            localStorage.themeProviderDebug = !isOpenState;
                            setIsOpenState(!isOpenState);
                        }}
                        size="sm"
                    >
                        ThemeProvider Debug
                    </Button>

                    <DebugScroll className="debug-scroll">
                        <Debug property={theme} propertyName="Theme" />
                    </DebugScroll>
                </DebugWrapper>
            }
            {children}
        </XstyledThemeProvider>
    );
};
