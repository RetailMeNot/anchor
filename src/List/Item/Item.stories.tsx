// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
// VENDOR
import styled from 'styled-components';
// COMPONENTS
import Item from './Item.component';

const StyledStory = styled.div`
  padding: 2rem 5rem;
`;

storiesOf('Components/Item', module)
  .add('Primary', () => (
      <StyledStory>
          <p>Item</p>
        <Item>Item 1</Item>
      </StyledStory>
  ));
