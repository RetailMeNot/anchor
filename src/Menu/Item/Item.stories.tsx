// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
// VENDOR
import styled, { ThemeProvider } from '@xstyled/styled-components';
// COMPONENTS
import { Item } from './Item.component';
import { RootTheme } from '../../theme';

const StyledStory = styled('div')`
    padding: 2rem 5rem;
`;

storiesOf('Components/Menu/Item', module).add('Primary', () => (
    <ThemeProvider theme={RootTheme}>
        <StyledStory>
            <p>Item</p>
            <Item label="Link" path="/" />
        </StyledStory>
    </ThemeProvider>
));
