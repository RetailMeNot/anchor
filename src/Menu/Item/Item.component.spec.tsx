// VENDOR
import * as React from 'react';
import * as renderer from 'react-test-renderer';
// import
import { ThemeProvider } from '@xstyled/styled-components';
// ANCHOR
import { RootTheme } from '../../theme';
import { Item } from './Item.component';

describe('Component: Item', () => {
    it('should be defined', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <Item path="/" />
            </ThemeProvider>
        );
        const tree = renderer.create(subject).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render with label', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <Item path="/" label="label" />
            </ThemeProvider>
        );

        const tree = renderer.create(subject).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
