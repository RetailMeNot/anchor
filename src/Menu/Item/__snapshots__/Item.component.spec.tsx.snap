// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Item should be defined 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  text-align: center;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #323232;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: color 250ms;
  transition: color 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-touch-callout: none;
}

<a
  className="c0 anchor-menu-item"
/>
`;

exports[`Component: Item should render with label 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  text-align: center;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  color: #323232;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: color 250ms;
  transition: color 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-touch-callout: none;
}

<a
  className="c0 anchor-menu-item"
>
  label
</a>
`;
